File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-4.cpp
#include <unordered_map>
#include <algorithm>
class Solution {
private:
    std::unordered_map<int,int> memo;
public:
    int minCostClimbingStairs(vector<int>& cost) {
        int res_1 = dfs(0,cost);
        memo.clear();
        int res_2 = dfs(1,cost);
        return std::min(res_1,res_2);
    }

    int dfs(int idx, const std::vector<int>& cost){
        if (memo.count(idx)==1){
            return memo[idx];
        }
        if (idx >= cost.size()){
            return 0;
        }

        memo[idx] = cost[idx] + std::min(dfs(idx+1,cost),dfs(idx+2,cost));
        return memo[idx];
    }

};

File: Data Structures & Algorithms/decode-ways/submission-4.py
class Solution:
    def numDecodings(self, s: str) -> int:
        memo = {}

        def dfs(s,i):
            print(s,i)
            if i in memo:
                return memo[i]
            if s == "":
                return 1
            if s[0] == "0":
                return 0
            if len(s) == 1:
                return 1

            if (s[0] == "1" and s[1] != "0") or (s[0] == "2" and s[1] !="0" and int(s[1])<7):
                memo[i] = dfs(s[1:],i+1) + dfs(s[2:],i+2)
            elif (s[0] == "1" and s[1] == "0") or ((s[0] == "2" and s[1] == "0")):
                memo[i] = dfs(s[2:],i+2)
            else:
                memo[i] = dfs(s[1:],i+1)
            
            return memo[i]
        
        return dfs(s,0)